File: plan-dashboard/scripts/visual-layout.mjs
/**
 * book.visual-layout/v1 — the human-curated visual placement contract (JS side).
 *
 * JS MIRROR of scripts/podcast/_visual_layout.py. The Astro "Book Composer"
 * writes book/visual-layout.json; render-book-pdf.mjs consumes it through this
 * module. Keep the enums, normalization, and schema string in sync with the
 * Python mirror in the same commit.
 *
 * Renderer semantics: flow=wrap -> float (align left|right) with text beside it,
 * requires width_pct<=50; flow=standalone -> centered full-column block;
 * align=center implies standalone. page_fit avoid|before|isolate-plate controls
 * pagination (never split / fresh page / own page).
 */
import { existsSync, readFileSync } from 'node:fs';
import path from 'node:path';

export const SCHEMA = 'book.visual-layout/v1';
export const ALIGNS = ['left', 'center', 'right'];
export const FLOWS = ['wrap', 'standalone'];
export const PAGE_FITS = ['avoid', 'before', 'isolate-plate'];
export const WRAP_MAX_WIDTH_PCT = 50;
export const DEFAULT_WIDTH_PCT = 60;

export function normalizePlacement(raw) {
  const warnings = [];
  const vid = String(raw.visual_id || '').trim();
  const anchor = String(raw.anchor || '').trim();

  let align = String(raw.align || 'center').trim().toLowerCase();
  if (!ALIGNS.includes(align)) { warnings.push(`${vid || '?'}: unknown align -> center`); align = 'center'; }

  let flow = String(raw.flow || 'standalone').trim().toLowerCase();
  if (!FLOWS.includes(flow)) { warnings.push(`${vid || '?'}: unknown flow -> standalone`); flow = 'standalone'; }

  let widthPct = parseInt(raw.width_pct, 10);
  if (!Number.isFinite(widthPct)) widthPct = DEFAULT_WIDTH_PCT;
  widthPct = Math.max(1, Math.min(100, widthPct));

  if (align === 'center' && flow === 'wrap') { warnings.push(`${vid}: center forces standalone`); flow = 'standalone'; }
  if (flow === 'wrap' && widthPct > WRAP_MAX_WIDTH_PCT) {
    warnings.push(`${vid}: wrap width ${widthPct}%>${WRAP_MAX_WIDTH_PCT}% -> standalone`);
    flow = 'standalone';
  }

  let pageFit = String(raw.page_fit || 'avoid').trim().toLowerCase();
  if (!PAGE_FITS.includes(pageFit)) { warnings.push(`${vid}: unknown page_fit -> avoid`); pageFit = 'avoid'; }

  return {
    placement: { visual_id: vid, anchor, align, flow, width_pct: widthPct, caption: String(raw.caption || '').trim(), page_fit: pageFit },
    warnings,
  };
}

export function validateLayout(data) {
  const warnings = [];
  if (!data || typeof data !== 'object') return { placements: [], warnings: ['layout is not an object'] };
  if (data.schema !== SCHEMA) warnings.push(`unexpected schema ${data.schema}`);
  const placements = [];
  for (const raw of data.placements || []) {
    if (!raw || typeof raw !== 'object' || !String(raw.visual_id || '').trim()) {
      warnings.push('placement without a visual_id skipped');
      continue;
    }
    const { placement, warnings: w } = normalizePlacement(raw);
    placements.push(placement);
    warnings.push(...w);
  }
  return { placements, warnings };
}

export function loadLayout(bookContentDir) {
  const p = path.join(bookContentDir, 'book', 'visual-layout.json');
  if (!existsSync(p)) return { placements: [], warnings: [] };
  try {
    return validateLayout(JSON.parse(readFileSync(p, 'utf-8')));
  } catch (e) {
    return { placements: [], warnings: [`visual-layout.json unreadable: ${e.message}`] };
  }
}

function esc(s) {
  return String(s).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
}

/**
 * Build the <figure> HTML for one placement. `assetSrc` is the URL the render
 * server can serve (e.g. /book/visuals/<file>). `embeddedTitle` (optional) is a
 * title already baked into the asset — when it equals the caption we suppress the
 * <figcaption> to avoid a duplicated caption (the original defect).
 * Width is data-driven, so it rides a CSS custom property on the figure; all
 * enumerable dimensions (flow/align/page_fit) are classes in book-print.css.
 */
export function figureHtml(placement, assetSrc, embeddedTitle = '') {
  const { align, flow, width_pct, caption, page_fit } = placement;
  const classes = ['v2-fig', `flow-${flow}`, `align-${align}`, `page-fit-${page_fit}`];
  // Both raster and SVG assets are served as files and embedded via <img>; the
  // render server serves them from book/visuals/ with the right MIME type.
  const inner = `<img src="${esc(assetSrc)}" alt="${esc(caption)}">`;
  const dupCaption = caption && embeddedTitle && caption.trim().toLowerCase() === embeddedTitle.trim().toLowerCase();
  const cap = caption && !dupCaption ? `<figcaption>${esc(caption)}</figcaption>` : '';
  return `<figure class="${classes.join(' ')}" style="--fig-w:${width_pct}%">${inner}${cap}</figure>`;
}

/**
 * Insert each placement's figure into `bodyHtml` immediately after its anchor
 * chapter <section class="chapter-open">…</section> block (the anchor is the
 * chapter heading text, e.g. "## 2. Title" or its rendered "2. Title").
 * `assetsById` maps visual_id -> { src, embeddedTitle }. Unknown ids are skipped
 * (tolerant of a partial contract). Returns the augmented body HTML.
 */
export function applyLayout(bodyHtml, placements, assetsById) {
  if (!placements.length) return bodyHtml;
  // Group figures by the anchor's chapter number/title so we can inject after
  // the chapter-open section. We match on the <h2> text the renderer emitted.
  const byAnchor = new Map();
  for (const p of placements) {
    const asset = assetsById.get(p.visual_id);
    if (!asset || !asset.src) continue;
    const key = anchorKey(p.anchor);
    if (!byAnchor.has(key)) byAnchor.set(key, []);
    byAnchor.get(key).push(figureHtml(p, asset.src, asset.embeddedTitle || ''));
  }
  if (!byAnchor.size) return bodyHtml;
  // Split on chapter-open sections; append matching figures right after each.
  const chapterRe = /(<section class="chapter-open[^"]*">[\s\S]*?<\/section>)/g;
  return bodyHtml.replace(chapterRe, (block) => {
    const h2 = block.match(/<h2>([\s\S]*?)<\/h2>/);
    if (!h2) return block;
    const key = anchorKey(h2[1]);
    const figs = byAnchor.get(key);
    return figs && figs.length ? `${block}\n${figs.join('\n')}` : block;
  });
}

/** Normalize an anchor / heading to a comparable key: strip markup, "N." prefix, case. */
export function anchorKey(s) {
  return String(s)
    .replace(/<[^>]+>/g, '')
    .replace(/^#{1,6}\s+/, '')
    .replace(/^\d+\.\s*/, '')
    .trim()
    .toLowerCase();
}

File: scripts/podcast/_visual_layout.py
"""book.visual-layout/v1 — the human-curated visual placement contract (Python side).

The Astro "Book Composer" (Phase 4) writes ``book/visual-layout.json``; the PDF
renderer (``plan-dashboard/scripts/visual-layout.mjs``) consumes it. This module
is the Python MIRROR of that JS validator — both sides must agree on the schema,
the allowed enum values, and the normalization rules, so keep them in sync in the
same commit (per the repo's TS<->Python mirror contract).

Renderer semantics the contract encodes:
  * ``flow: wrap``  -> the image floats (``align: left|right``) with body text
    flowing beside it. Requires ``width_pct <= 50`` so the text has a real column;
    a wider wrap is coerced to ``standalone`` (a too-wide float leaves a sliver of
    text and reads as broken).
  * ``flow: standalone`` -> a full-column centered block. ``align: center`` always
    implies standalone.
  * ``page_fit: avoid`` -> never split the figure across a page break (default).
    ``before`` -> start the figure on a fresh page. ``isolate-plate`` -> give the
    figure its own page (used for a tall plate that would otherwise strand text).
"""
from __future__ import annotations

from pathlib import Path
from typing import Any

SCHEMA = "book.visual-layout/v1"

ALIGNS = ("left", "center", "right")
FLOWS = ("wrap", "standalone")
PAGE_FITS = ("avoid", "before", "isolate-plate")

WRAP_MAX_WIDTH_PCT = 50  # a float must leave a real text column beside it
DEFAULT_WIDTH_PCT = 60


def normalize_placement(raw: dict[str, Any]) -> tuple[dict[str, Any], list[str]]:
    """Return (normalized placement, warnings). Never raises on a single bad field."""
    warnings: list[str] = []
    vid = str(raw.get("visual_id") or "").strip()
    anchor = str(raw.get("anchor") or "").strip()

    align = str(raw.get("align") or "center").strip().lower()
    if align not in ALIGNS:
        warnings.append(f"{vid or '?'}: unknown align {align!r} -> center")
        align = "center"

    flow = str(raw.get("flow") or "standalone").strip().lower()
    if flow not in FLOWS:
        warnings.append(f"{vid or '?'}: unknown flow {flow!r} -> standalone")
        flow = "standalone"

    try:
        width_pct = int(raw.get("width_pct") or DEFAULT_WIDTH_PCT)
    except (TypeError, ValueError):
        width_pct = DEFAULT_WIDTH_PCT
    width_pct = max(1, min(100, width_pct))

    # align:center is centered => cannot float.
    if align == "center" and flow == "wrap":
        warnings.append(f"{vid or '?'}: align=center forces flow=standalone")
        flow = "standalone"
    # A wrap wider than the cap has no real text column -> promote to standalone.
    if flow == "wrap" and width_pct > WRAP_MAX_WIDTH_PCT:
        warnings.append(
            f"{vid or '?'}: wrap width {width_pct}%>{WRAP_MAX_WIDTH_PCT}% -> standalone"
        )
        flow = "standalone"

    page_fit = str(raw.get("page_fit") or "avoid").strip().lower()
    if page_fit not in PAGE_FITS:
        warnings.append(f"{vid or '?'}: unknown page_fit {page_fit!r} -> avoid")
        page_fit = "avoid"

    return (
        {
            "visual_id": vid,
            "anchor": anchor,
            "align": align,
            "flow": flow,
            "width_pct": width_pct,
            "caption": str(raw.get("caption") or "").strip(),
            "page_fit": page_fit,
        },
        warnings,
    )


def validate_layout(data: dict[str, Any]) -> tuple[list[dict[str, Any]], list[str]]:
    """Validate + normalize a whole layout. Returns (placements, warnings).

    Tolerant by design: a malformed schema header or a bad placement produces a
    warning, not an exception — a partial contract must still render.
    """
    warnings: list[str] = []
    if not isinstance(data, dict):
        return [], ["layout is not an object"]
    if data.get("schema") != SCHEMA:
        warnings.append(f"unexpected schema {data.get('schema')!r} (expected {SCHEMA})")
    placements: list[dict[str, Any]] = []
    for raw in data.get("placements") or []:
        if not isinstance(raw, dict) or not str(raw.get("visual_id") or "").strip():
            warnings.append("placement without a visual_id skipped")
            continue
        norm, w = normalize_placement(raw)
        placements.append(norm)
        warnings.extend(w)
    return placements, warnings


def load_layout(book_dir: Path) -> tuple[list[dict[str, Any]], list[str]]:
    """Load book/visual-layout.json if present. Absent contract => ([], [])."""
    import json

    path = Path(book_dir) / "book" / "visual-layout.json"
    if not path.exists():
        return [], []
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except Exception as e:  # noqa: BLE001
        return [], [f"visual-layout.json unreadable: {e}"]
    return validate_layout(data)
